design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,230 @@
+pc-q35-rhel9.2.0     RHEL-9.2.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel9.0.0     RHEL-9.0.0 PC (Q35 + ICH9, 2009)
+pc-q35-rhel8.6.0     RHEL-8.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.5.0     RHEL-8.5.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.4.0     RHEL-8.4.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.3.0     RHEL-8.3.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.2.0     RHEL-8.2.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.1.0     RHEL-8.1.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel8.0.0     RHEL-8.0.0 PC (Q35 + ICH9, 2009) (deprecated)
+pc-q35-rhel7.6.0     RHEL-7.6.0 PC (Q35 + ICH9, 2009) (deprecated)
+none                 empty machine
+This means each workload must be updated to use a `pc-q35-rhel9.x.x` machine type.
+````
+## Goals
+* Create an opt-in workflow that automates mass converting machine types of VMs
+	* if currently running, update VM machine type and add a label to alert the user the VM must be restarted for the change to take effect
+	* if currently running, take VM offline immediately and update machine type
+	* if not currently running, update machine type immediately
+* Allow user to specify certain VMs for updating machine type automation (e.g. by namespace)
+
+## Non Goals
+While these additions could be beneficial to cluster-admins who wish to use the mass machine type transition (MMTT), these are not covered in the scope of this design proposal: 
+* Allowing user to filter VMs by running state: a cluster admin may want to only update VMs that are offline.
+* Allowing user to filter VMs in ways other than by namespace; e.g. with label-selector
+
+Also not covered in this design proposal is the method in which users/cluster admins will be informed that they have VMs with unsupported machine types.
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with CentOS Stream 10 without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. Both the minimum supported machine type and the target machine type are constants that can be updated as future versions are released and more machine types are deprecated upon these releases. 
```

Comment:
Is there a benefit in running a `Job` instead of doing it directly in `virtctl`? I guess it is about run time of the command vs. having a separate binary to run inside the `Job`?

Reply:
Hey @0xFelix!

As I see it, the main benefit of running this inside a job is that you can get a status of the mass conversion process. For example, you can see that you had 20 VMs that were outdated, but now 7 had restarted so there are 13 more to go, the remaining VM names, etc. In addition, you can clearly see when the whole operation succeeds, meaning that you're sure your cluster is fully updated. A job would also automatically add/remove warning labels.

Doing so directly with virtctl would put the burden of tracking the outdated VMs, removing warning labels, and tracking the status of the overall conversion task on the user.

Essentially, this approach is more declarative than imperative. Since as an admin I would like to declaratively state "I want all of the outdated VMs to convert their machine type into a new one", and not mind about how exactly this is being done behind the scenes, I think a job fits perfectly here.

WDYT?

Reply:
I was just concerned about the run time of the Job and its state. Say the cluster has a large amount of VMs and the Job may gets killed inbetween. How does the user know? Will there be any state left-overs from the Job? Can it be restarted to resume the operation?

Also, how often would this Job be needed? How often do machine types change?

Reply:
IOW it may become a bit imperative, as the users needs to check on the job to make sure it succeeded or to restart it. Declarative would be more like declare it once and forget about it?

Reply:
Thanks a lot @0xFelix for this discussion!

> Would be good if the two points were added to the proposal.

I completely agree.
@prnaraya WDYT?

Reply:
I've updated the design proposal to include how the Job will be handled if it is killed and how leftovers (leftover labels) will be handled upon the Job restarting.